README.md
# @daeinc/timeline

A Timeline object can hold multiple properties with keyframes. ex. position, rotation, color, etc., and makes it easy to group things together. In my own projects, I create multiple Timeline objects, one for each graphical object.

It builds on top of [`keyframes`](https://github.com/mattdesl/keyframes), but does not expose Keyframes library to user - user only needs to provide a minimum amount of necessary data. ex. `{ time, value, ... }`

In development, and breaking changes are expected in the future.

## Installation

```
npm i @daeinc/timeline
```

Then,

```js
import Timeline from "@daeinc/timeline";
```

## Examples

Pass a timeline name and a single property object (`{name, keyframes}`) or an array of property objects (`[{name, keyframes}, {name, keyframes}]`):

```js
const tl = new Timeline("my-timeline", {
  name: "position",
  keyframes: [
    { time: 0, value: 5 },
    { time: 1, value: 10 },
  ],
});

const v1 = tl.value("position", 0.5); // returns 7.5
```

Pass a custom interpolator function:

JS:

```js
const easeInQuad = (a, b, t) => {
  return lerp(a.value, b.value, t * t);
};

const v2 = tl.value("position", 0.5, easeInQuad);
console.log(v2); // 6.25
```

TS:

```ts
import type { Keyframe } from "@daeinc/timeline";

const easeInQuad = (a: Keyframe, b: Keyframe, t: number) => {
  return lerp(a.value, b.value, t * t);
};

const v2 = tl.value("position", 0.5, easeInQuad);
console.log(v2); // 6.25
```

## Methods

```ts
constructor(propOrProps?: InputProp | InputProp[]);

propExists(propName: string): boolean;

addKeyframe(propName: string, newKey: Keyframe): void;

addKeyframes(propName: string, ...newKeys: Keyframe[]): void;

getKeyframe(propName: string, timeStamp: number): Keyframe;

getKeyframes(propName: string): Keyframe[];

getName(): string;

getPropertyNames(): string[];

value(propName: string, timeStamp: number, interpolator?: Interpolator): any;

addProperty(propName: string, ...newKeys: Keyframe[]): void;

removeKeyframes(propName: string, index: number, howmany: number): void;

nearest(propName: string, timeStamp: number, radius?: number): Keyframe;

nearestIndex(propName: string, timeStamp: number, radius?: number): number;

next(propName: string, timeStamp: number): Keyframe;

previous(propName: string, timeStamp: number): Keyframe;

static fromJSON(file: string): Promise<void | Timeline>;

static from(data: {
  name: string;
  properties: InputProp[];
}): Timeline;

toJSON(): string;
```

## To Dos

- add more examples
  - how to add keyframes
  - how to import and export
- implement the rest of the methods from `keyframes` package
- add a separate `.d.ts` for `keyframes` JS package, or, create a TS version of the keyframes package as a `class`.
- throw errors or return null or guard against them?

## License

MIT


CHANGELOG.md
# Changelog

## `0.4.0`

updated build to `esnext`

## `0.3.0`

added `fromJSON()`, `toJSON()`, `from()` methods

## `0.2.0`

added `name` field to Timeline

## `0.1.0`

first release


index.test.ts
const { describe, test, expect } = require("@jest/globals");
import Timeline from "./index";
import type { Interpolator } from "./index";
import { mix } from "@daeinc/math";

const prop1 = { name: "position", keyframes: [] };
const prop2 = {
  name: "position",
  keyframes: [{ time: 0, value: 10 }],
};
const prop3 = {
  name: "position",
  keyframes: [
    { time: 0, value: 10 },
    { time: 1, value: 20 },
  ],
};
const prop4 = {
  name: "color",
  keyframes: [
    { time: 0, value: [0, 0, 0] },
    { time: 1, value: [0, 0, 1] },
  ],
};

const name = "rect-1";

// with no properties
const tl1 = new Timeline(name);
const tl2 = new Timeline(name, []);

// single property object
const tl3 = new Timeline(name, prop1);
const tl4 = new Timeline(name, prop3);

// with properties array + keyframes
const tl5 = new Timeline(name, [prop1]);
const tl6 = new Timeline(name, [prop2]);
const tl7 = new Timeline(name, [prop3]);
const tl8 = new Timeline(name, [prop3, prop4]);

const linear: Interpolator = (a: any, b: any, t: number) =>
  mix(a.value, b.value, t);

describe("Timeline constructor()", () => {
  test("creates a new Timeline object", () => {
    expect(typeof tl1).toEqual("object");
    expect(typeof tl2.value).toEqual("function");
    expect(typeof tl6.value).toEqual("function");
    expect(Array.isArray(tl1.properties)).toEqual(true);
    expect(Array.isArray(tl2.properties)).toEqual(true);
    expect(Array.isArray(tl3.properties)).toEqual(true);
    expect(Array.isArray(tl5.properties)).toEqual(true);
  });
  test("can access properties array", () => {
    expect(tl3.properties.length).toEqual(1);
    expect(tl8.properties.length).toEqual(2);
    expect(typeof tl3.properties[0]).toEqual("object");
    expect(tl3.properties[0].name).toEqual("position");
  });
  test("creates a new object without argument", () => {
    expect(tl1 instanceof Timeline).toEqual(true);
    expect(tl2 instanceof Timeline).toEqual(true);
  });
});

describe("propExists()", () => {
  test("returns false if property does not exist", () => {
    expect(tl1.propExists("position")).toEqual(false);
  });
  test("returns true if property exists", () => {
    expect(tl3.propExists("position")).toEqual(true);
    expect(tl8.propExists("color")).toEqual(true);
  });
});

describe("addKeyframe()", () => {
  test("adds a new keyframe and a new property", () => {
    tl1.addKeyframe("opacity", { time: 0, value: 0.5 });
    tl2.addKeyframe("opacity", { time: 0, value: 0.5 });
    expect(tl1.propExists("opacity")).toEqual(true);
    expect(tl1.properties.length).toEqual(1);
    expect(tl1.properties[0].name).toEqual("opacity");
    expect(tl1.properties[0].keyframes.frames[0].value).toEqual(0.5);
    expect(tl2.propExists("opacity")).toEqual(true);
    expect(tl2.properties.length).toEqual(1);
    expect(tl2.properties[0].name).toEqual("opacity");
    expect(tl2.properties[0].keyframes.frames[0].value).toEqual(0.5);
  });
  test("adds a new keyframe to an existing property", () => {
    const tl = new Timeline(name, {
      name: "position",
      keyframes: [{ time: 1, value: 5 }],
    });
    tl.addKeyframe("position", { time: 2, value: 10 });
    expect(tl.properties[0].name).toBe("position");
    expect(tl.properties[0].keyframes.frames[1].time).toBe(2);
    expect(tl.properties[0].keyframes.frames[1].value).toBe(10);
  });
  test("replaces a keyframe if timeStamp is the same", () => {
    const tl = new Timeline(name, {
      name: "position",
      keyframes: [{ time: 1, value: 5 }],
    });
    tl.addKeyframe("position", { time: 1, value: 10 });
    expect(tl.properties[0].name).toBe("position");
    expect(tl.properties[0].keyframes.frames.length).toBe(1);
    expect(tl.properties[0].keyframes.frames[0].value).toBe(10);
  });
  test("handles an optional filed (ease) in a keyframe", () => {
    const tl = new Timeline(name, {
      name: "position",
      keyframes: [{ time: 1, value: 5 }],
    });
    tl.addKeyframe("position", { time: 2, value: 10, ease: "sineIn" });
    expect(tl.properties[0].keyframes.frames[1].value).toBe(10);
    expect(tl.properties[0].keyframes.frames[1].ease).toBe("sineIn");
  });
});

describe("addKeyframes()", () => {
  test("adds multiple new keyframes in one call", () => {
    const tl = new Timeline(name);
    tl.addKeyframes("position", { time: 0, value: 1 }, { time: 1, value: 10 });
    expect(tl.properties[0].name).toEqual("position");
    expect(tl.properties[0].keyframes.frames.length).toEqual(2);
    expect(tl.properties[0].keyframes.frames[1].value).toEqual(10);
  });
});

describe("getKeyframe()", () => {
  const tl = new Timeline(name, {
    name: "position",
    keyframes: [
      { time: 1, value: 5 },
      { time: 2, value: 10 },
    ],
  });
  test("throws error if a prop does not exist", () => {
    expect(() => tl.getKeyframe("color", 1)).toThrow("cannot find prop");
  });
  test("throws error if a key does not exist", () => {
    expect(() => tl.getKeyframe("position", 0.8)).toThrow("key does not exist");
  });
  test("returns a keyframe object", () => {
    const key = tl.getKeyframe("position", 1);
    expect(key).toEqual({ time: 1, value: 5 });
  });
});

describe("getKeyframes()", () => {
  test("returns keyframes object", () => {
    const tl = new Timeline(name, {
      name: "position",
      keyframes: [],
    });
    const tl2 = new Timeline(name, {
      name: "position",
      keyframes: [
        { time: 1, value: 5 },
        { time: 2, value: 10 },
      ],
    });
    expect(tl.getKeyframes("position")).toEqual([]);
    expect(tl2.getKeyframes("position")).toEqual([
      { time: 1, value: 5 },
      { time: 2, value: 10 },
    ]);
  });
});

describe("getName()", () => {
  test("returns name of timeline", () => {
    expect(tl1.getName()).toEqual("rect-1");
  });
});

describe("getPropertyNames()", () => {
  test("returns name of all properties", () => {
    expect(tl8.getPropertyNames()).toEqual(["position", "color"]);
  });
});

describe("value()", () => {
  const tl = new Timeline(name, {
    name: "position",
    keyframes: [
      { time: 0, value: 5 },
      { time: 1, value: 10 },
    ],
  });
  test("returns correct value", () => {
    expect(tl.value("position", 0)).toEqual(5);
    expect(tl.value("position", 0.5)).toEqual(7.5);
    expect(tl.value("position", 1)).toEqual(10);
  });
  test("returns correct value with interpolator", () => {
    expect(tl.value("position", 0, linear)).toEqual(5);
    expect(tl.value("position", 0.5, linear)).toEqual(7.5);
  });
  test("returns clamped value if outside timeStamp range", () => {
    expect(tl.value("position", 2, linear)).toEqual(10);
  });
  test("interpolates array value", () => {
    // TODO
  });
});

describe("addProperty()", () => {
  const tl = new Timeline(name, {
    name: "position",
    keyframes: [
      { time: 0, value: 5 },
      { time: 1, value: 10 },
    ],
  });
  test("adds a new property without keyframe", () => {
    tl.addProperty("rotation");
    expect(tl.properties[1].name).toEqual("rotation");
    expect(tl.properties[1].keyframes.frames).toEqual([]);
  });
  test("adds a new property with keyframes", () => {
    tl.addProperty("alpha", { time: 1, value: 0.5 });
    tl.addProperty(
      "color",
      { time: 0, value: 0 },
      { time: 1, value: 1 },
      { time: 2, value: 2 }
    );
    expect(tl.properties[2].name).toEqual("alpha");
    expect(tl.properties[2].keyframes.frames.length).toEqual(1);
    expect(tl.properties[3].name).toEqual("color");
    expect(tl.properties[3].keyframes.frames.length).toEqual(3);
    expect(tl.properties[3].keyframes.frames[2]).toEqual({ time: 2, value: 2 });
  });
  test("handles when property already exists", () => {
    expect(() => tl.addProperty("position")).toThrow("already exists");
    expect(() => tl.addProperty("position", { time: 2, value: 15 })).toThrow(
      "already exists"
    );
  });
});

describe("removeKeyframes()", () => {
  test("removes single keyframe", () => {
    const tl = new Timeline(name, {
      name: "position",
      keyframes: [
        { time: 0, value: 5 },
        { time: 1, value: 10 },
        { time: 2, value: 15 },
      ],
    });
    tl.removeKeyframes("position", 1, 1);
    expect(tl.properties[0].name).toEqual("position");
    expect(tl.properties[0].keyframes.frames.length).toEqual(2);
    expect(tl.properties[0].keyframes.frames[1]).toEqual({
      time: 2,
      value: 15,
    });
  });
  test("removes multiple keyframes", () => {
    const tl = new Timeline(name, {
      name: "position",
      keyframes: [
        { time: 0, value: 5 },
        { time: 1, value: 10 },
        { time: 2, value: 15 },
      ],
    });
    tl.removeKeyframes("position", 0, 2);
    expect(tl.properties[0].name).toEqual("position");
    expect(tl.properties[0].keyframes.frames.length).toEqual(1);
    expect(tl.properties[0].keyframes.frames[0]).toEqual({
      time: 2,
      value: 15,
    });
  });
  test("returns empty [] when keyframes are empty", () => {
    const tl = new Timeline(name, {
      name: "position",
      keyframes: [{ time: 0, value: 5 }],
    });
    tl.removeKeyframes("position", 0, 2);
    expect(tl.properties[0].name).toEqual("position");
    expect(tl.properties[0].keyframes.frames).toEqual([]);
  });
});

describe("nearest()", () => {
  const tl = new Timeline(name, {
    name: "position",
    keyframes: [
      { time: 0, value: 5 },
      { time: 1, value: 10 },
      { time: 2, value: 15 },
    ],
  });
  test("returns nearest keyframe object", () => {
    expect(tl.nearest("position", 0.2, 0.5)).toEqual({ time: 0, value: 5 });
  });
  test("clamps to either end of keyframe object", () => {
    expect(tl.nearest("position", -0.5, 0.5)).toEqual({ time: 0, value: 5 });
    expect(tl.nearest("position", 2.5, 0.5)).toEqual({ time: 2, value: 15 });
  });
});

describe("nearestIndex()", () => {
  const tl = new Timeline(name, {
    name: "position",
    keyframes: [
      { time: 0, value: 5 },
      { time: 1, value: 10 },
      { time: 2, value: 15 },
    ],
  });
  test("returns nearest index", () => {
    expect(tl.nearestIndex("position", 0.4, 0.5)).toEqual(0);
  });
  test("clamps to either end of nearest index", () => {
    expect(tl.nearestIndex("position", -0.5)).toEqual(0);
    expect(tl.nearestIndex("position", 2.5, 0.5)).toEqual(2);
  });
});
describe("next()", () => {
  const tl = new Timeline(name, {
    name: "position",
    keyframes: [
      { time: 0, value: 5 },
      { time: 1, value: 10 },
      { time: 2, value: 15 },
    ],
  });
  test("returns next keyframe object", () => {
    expect(tl.next("position", 0.1)).toEqual({ time: 1, value: 10 });
    expect(tl.next("position", 1.4)).toEqual({ time: 2, value: 15 });
  });
  test("clamps to previous keyframe object", () => {
    expect(tl.next("position", 12.5)).toEqual({ time: 2, value: 15 });
  });
});
describe("previous()", () => {
  const tl = new Timeline(name, {
    name: "position",
    keyframes: [
      { time: 0, value: 5 },
      { time: 1, value: 10 },
      { time: 2, value: 15 },
    ],
  });
  test("returns previous keyframe object", () => {
    expect(tl.previous("position", 0.1)).toEqual({ time: 0, value: 5 });
    expect(tl.previous("position", 1.4)).toEqual({ time: 1, value: 10 });
  });
  test("clamps to next keyframe object", () => {
    expect(tl.previous("position", -1.5)).toEqual({ time: 0, value: 5 });
  });
});

// describe("get()", () => {
//   const tl = new Timeline(name, {
//     name: "position",
//     keyframes: [
//       { time: 0, value: 5 },
//       { time: 1, value: 10 },
//     ],
//   });
//   test("returns keyframe object", () => {
//     expect(tl.get("position", 0)).toEqual({ time: 0, value: 5 });
//   });
//   test("returns closest keyframe object", () => {
//     expect(tl.get("position", 0.4)).toEqual({ time: 0, value: 5 });
//   });
// });
// describe("getIndex()", () => {
//   test("", () => {
//     // expect().toEqual()
//   });
// });

describe("fromJSON()", () => {
  test("creates a new Timeline object from JSON file", () => {
    // TODO
    // expect().toEqual()
  });
});

describe("toJSON()", () => {
  test("convert data to JSON format", () => {
    // TODO
    // expect().toEqual()
  });
});

describe("from()", () => {
  test("create a new Timeline object from JS data object", () => {
    // TODO
    // expect().toEqual()
  });
});


ease.js
/**
 * LICENSE
 * - equations: https://github.com/ai/easings.net/blob/master/LICENSE
 *
 * TOFIX:
 * - switch/case and return func() seems a bit strange...
 * @param {number} x 0..1
 * @param {string} type type of easing
 * @returns eased number
 */
function ease(x, type) {
  switch (type) {
    case "sineInOut":
      return sineInOut(x);
    case "quadIn":
      return quadIn(x);
    case "quadOut":
      return quadOut(x);
    case "quadInOut":
      return quadInOut(x);
    // case "easeOutInQuad":
    //   return x < 0.5
    //     ? 0.5 - Math.pow(1 - 2 * x, 2) * 0.5
    //     : Math.pow(x - 0.5, 2) * 2 + 0.5;
    case "cubicIn":
      return cubicIn(x);
    case "cubicOut":
      return cubicOut(x);
    case "cubicInOut":
      return cubicInOut(x);
    case "quartIn":
      return quartIn(x);
    case "quartOut":
      return quartOut(x);
    case "quartInOut":
      return quartInOut(x);
    case "quintIn":
      return quintIn(x);
    case "quintOut":
      return quintOut(x);
    case "quintInOut":
      return quintInOut(x);
    case "expoIn":
      return expoIn(x);
    case "expoOut":
      return expoOut(x);
    case "expoInOut":
      return expoInOut(x);
    case "backIn":
      return backIn(x);
    case "backOut":
      return backOut(x);
    case "backInOut":
      return backInOut(x);
    case "bounceIn":
      return bounceIn(x);
    case "bounceOut":
      return bounceOut(x);
    case "bounceInOut":
      return bounceInOut(x);
    case "linear":
    default:
      return x;
  }
}

// TODO: rest of sine and others
function sineInOut(x) {
  return -(Math.cos(Math.PI * x) - 1) / 2;
}

function quadIn(x) {
  return x * x;
}
function quadOut(x) {
  return 1 - (1 - x) * (1 - x);
}
function quadInOut(x) {
  return x < 0.5 ? 2 * x * x : 1 - Math.pow(-2 * x + 2, 2) / 2;
}
function cubicIn(x) {
  return x * x * x;
}
function cubicOut(x) {
  return 1 - Math.pow(1 - x, 3);
}
function cubicInOut(x) {
  return x < 0.5 ? 4 * x * x * x : 1 - Math.pow(-2 * x + 2, 3) / 2;
}
function quartIn(x) {
  return Math.pow(x, 4);
}
function quartOut(x) {
  return 1 - Math.pow(1 - x, 4);
}
function quartInOut(x) {
  return x < 0.5 ? 8 * x * x * x * x : 1 - Math.pow(-2 * x + 2, 4) / 2;
}
function quintIn(x) {
  return Math.pow(x, 5);
}
function quintOut(x) {
  return 1 - Math.pow(1 - x, 5);
}
function quintInOut(x) {
  return x < 0.5 ? 16 * Math.pow(x, 5) : 1 - Math.pow(-2 * x + 2, 5) / 2;
}
function expoIn(x) {
  return x === 0 ? 0 : Math.pow(2, 10 * x - 10);
}
function expoOut(x) {
  return x === 1 ? 1 : 1 - Math.pow(2, -10 * x);
}
function expoInOut(x) {
  return x === 0
    ? 0
    : x === 1
    ? 1
    : x < 0.5
    ? Math.pow(2, 20 * x - 10) / 2
    : (2 - Math.pow(2, -20 * x + 10)) / 2;
}
function backIn(x) {
  const eibc1 = 1.70158;
  const eibc3 = eibc1 + 1;
  return eibc3 * x * x * x - eibc1 * x * x;
}
function backOut(x) {
  const eobc1 = 1.70158;
  const eobc3 = eobc1 + 1;
  return 1 + eobc3 * Math.pow(x - 1, 3) + eobc1 * Math.pow(x - 1, 2);
}
function backInOut(x) {
  const eiobc1 = 1.70158;
  const eiobc2 = eiobc1 * 1.525;
  return x < 0.5
    ? (Math.pow(2 * x, 2) * ((eiobc2 + 1) * 2 * x - eiobc2)) / 2
    : (Math.pow(2 * x - 2, 2) * ((eiobc2 + 1) * (x * 2 - 2) + eiobc2) + 2) / 2;
}
function bounceIn(x) {
  return 1 - bounceOut(1 - x);
}
function bounceOut(x) {
  const n1 = 7.5625;
  const d1 = 2.75;
  if (x < 1 / d1) {
    return n1 * x * x;
  } else if (x < 2 / d1) {
    return n1 * (x -= 1.5 / d1) * x + 0.75;
  } else if (x < 2.5 / d1) {
    return n1 * (x -= 2.25 / d1) * x + 0.9375;
  } else {
    return n1 * (x -= 2.625 / d1) * x + 0.984375;
  }
}
function bounceInOut(x) {
  return x < 0.5
    ? (1 - bounceOut(1 - 2 * x)) / 2
    : (1 + bounceOut(2 * x - 1)) / 2;
}

module.exports = {
  default: ease,
  ease,
  // TODO: sineIn, sineOut, and the rest...
  sineInOut,
  quadIn,
  quadOut,
  quadInOut,
  cubicIn,
  cubicOut,
  cubicInOut,
  quartIn,
  quartOut,
  quartInOut,
  quintIn,
  quintOut,
  quintInOut,
  expoIn,
  expoOut,
  expoInOut,
  backIn,
  backOut,
  backInOut,
  bounceIn,
  bounceOut,
  bounceInOut,
};


demo/data.js
/**
 * NOTES:
 * - naming conventions:
 *   - entry name: object-numbering. ex. "circle-1", "rect-3" (may be parsed later.)
 *   - property name:; follow Canvas API as much as possible.
 *   - ease: can have "hold" value -> this needs to be handled when *applying* easing in the main code.
 *
 * Data structure:
 * timeline (array)
 *   [entry] (object)
 *     name (string)
 *     properties (array)
 *       [property] (object)
 *         name (string)
 *         keyframes (array => later, converted to Keyframes in Timeline object)
 *
 */

const data = [
  {
    name: "circle-1",
    properties: [
      {
        name: "position",
        keyframes: [
          { time: 0, value: [500, 100], ease: "hold" },
          { time: 1, value: [300, 300], ease: "expoIn" },
          { time: 2, value: [300, 500], ease: "quadInOut" },
        ],
      },
      {
        name: "lineWidth",
        keyframes: [
          { time: 0, value: 1 },
          { time: 2, value: 20 },
          { time: 3, value: 4 },
        ],
      },
      {
        name: "strokeColor",
        keyframes: [
          { time: 0, value: [0.7, 1, 1] },
          { time: 1, value: [0, 1, 1], ease: "expoIn" },
          { time: 2, value: [0.2, 1, 1], ease: "expoIn" },
        ],
      },
    ],
  },
  {
    name: "rect-1",
    properties: [
      {
        name: "position",
        keyframes: [
          { time: 0, value: [100, 300] },
          { time: 1.5, value: [400, 400], ease: "expoIn" },
          { time: 3, value: [100, 300], ease: "expoOut" },
        ],
      },
    ],
  },
  {
    name: "shape-1",
    properties: [
      {
        name: "path",
        keyframes: [
          {
            time: 0,
            value: [
              [50, 50],
              [520, 50],
              [400, 400],
            ],
          },
          {
            time: 1.5,
            value: [
              [80, 150],
              [520, 180],
              [290, 540],
            ],
            ease: "expoInOut",
          },
          {
            time: 3,
            value: [
              [50, 50],
              [520, 50],
              [400, 400],
            ],
            ease: "expoInOut",
          },
        ],
      },
      {
        name: "lineWidth",
        keyframes: [
          { time: 0, value: 8 },
          { time: 1, value: 40 },
          { time: 2, value: 8 },
        ],
      },
      {
        name: "strokeColor",
        keyframes: [
          { time: 0, value: [0, 0, 1] },
          { time: 1.5, value: [0.4, 1, 1], ease: "expoInOut" },
          { time: 3, value: [0, 0, 1], ease: "expoInOut" },
        ],
      },
    ],
  },
];

export default data;


demo/demo.ts
import Timeline from "../index";
import { lerp } from "@daeinc/math";
import { interpolateArray } from "@daeinc/array";
import type { Keyframe } from "../index";
import { single, multiple } from "./data";

const tl = new Timeline("my-timeline", {
  name: "position",
  keyframes: [
    { time: 0, value: 5 },
    { time: 1, value: 10 },
  ],
});

const v1 = tl.value("position", 0.5);
console.log(v1); // 7.5

const easeInQuad = (a: Keyframe, b: Keyframe, t: number) => {
  return lerp(a.value, b.value, t * t);
};

const v2 = tl.value("position", 0.5, easeInQuad);
console.log(v2); // 6.25

// this is a workaround due to top-level await error.
const main = async () => {
  const tl2 = (await Timeline.fromJSON("./single.json")) as Timeline;
  const val = tl2.value("position", 0.5);
  console.log("JSON to Timeline", val);

  const backToJson = tl2.toJSON();
  console.log(backToJson)
  const tl3 = Timeline.from(JSON.parse(backToJson));
  const v3 = tl3.value("position", 0.5);
  console.log("JSON to Timeline to JSON to Timeline", v3);
};
main();

const tl4 = Timeline.from(single);
const v4 = tl4.value("position", 0.5);
console.log("Timeline.from().value()", v4);

// TODO: handle importing JSON with multiple timelines


demo/data.ts
/**
 * NOTES:
 * - naming conventions:
 *   - entry name: object-numbering. ex. "circle-1", "rect-3" (may be parsed later.)
 *   - property name:; follow Canvas API as much as possible.
 *   - ease: can have "hold" value -> this needs to be handled when *applying* easing in the main code.
 *
 * Data structure:
 * timeline (array)
 *   [entry] (object)
 *     name (string)
 *     properties (array)
 *       [property] (object)
 *         name (string)
 *         keyframes (array => later, converted to Keyframes in Timeline object)
 *
 */

export const single = {
  name: "circle-1",
  properties: [
    {
      name: "position",
      keyframes: [
        { time: 0, value: [500, 100], ease: "hold" },
        { time: 1, value: [300, 300], ease: "expoIn" },
        { time: 2, value: [300, 500], ease: "quadInOut" },
      ],
    },
    {
      name: "lineWidth",
      keyframes: [
        { time: 0, value: 1 },
        { time: 2, value: 20 },
        { time: 3, value: 4 },
      ],
    },
    {
      name: "strokeColor",
      keyframes: [
        { time: 0, value: [0.7, 1, 1] },
        { time: 1, value: [0, 1, 1], ease: "expoIn" },
        { time: 2, value: [0.2, 1, 1], ease: "expoIn" },
      ],
    },
  ],
};

export const multiple = [
  {
    name: "circle-1",
    properties: [
      {
        name: "position",
        keyframes: [
          { time: 0, value: [500, 100], ease: "hold" },
          { time: 1, value: [300, 300], ease: "expoIn" },
          { time: 2, value: [300, 500], ease: "quadInOut" },
        ],
      },
      {
        name: "lineWidth",
        keyframes: [
          { time: 0, value: 1 },
          { time: 2, value: 20 },
          { time: 3, value: 4 },
        ],
      },
      {
        name: "strokeColor",
        keyframes: [
          { time: 0, value: [0.7, 1, 1] },
          { time: 1, value: [0, 1, 1], ease: "expoIn" },
          { time: 2, value: [0.2, 1, 1], ease: "expoIn" },
        ],
      },
    ],
  },
  {
    name: "rect-1",
    properties: [
      {
        name: "position",
        keyframes: [
          { time: 0, value: [100, 300] },
          { time: 1.5, value: [400, 400], ease: "expoIn" },
          { time: 3, value: [100, 300], ease: "expoOut" },
        ],
      },
    ],
  },
  {
    name: "shape-1",
    properties: [
      {
        name: "path",
        keyframes: [
          {
            time: 0,
            value: [
              [50, 50],
              [520, 50],
              [400, 400],
            ],
          },
          {
            time: 1.5,
            value: [
              [80, 150],
              [520, 180],
              [290, 540],
            ],
            ease: "expoInOut",
          },
          {
            time: 3,
            value: [
              [50, 50],
              [520, 50],
              [400, 400],
            ],
            ease: "expoInOut",
          },
        ],
      },
      {
        name: "lineWidth",
        keyframes: [
          { time: 0, value: 8 },
          { time: 1, value: 40 },
          { time: 2, value: 8 },
        ],
      },
      {
        name: "strokeColor",
        keyframes: [
          { time: 0, value: [0, 0, 1] },
          { time: 1.5, value: [0.4, 1, 1], ease: "expoInOut" },
          { time: 3, value: [0, 0, 1], ease: "expoInOut" },
        ],
      },
    ],
  },
];
